File: 0291-Word Pattern II/0291-Word Pattern II.cpp
class Solution {
public:
    bool wordPatternMatch(string pattern, string str) {
        return dfs(pattern, 0, str, 0);
    }
    
private:
    unordered_map<string, char> table;
    unordered_set<char> chars;
    bool dfs(string& pattern, int pstart, string& str, int sstart)
    {
        if (pstart == pattern.size())
        {
            if (sstart == str.size())
            {
                return true;
            }
            else
            {
                return false;
            }
        }
        
        for (int i = sstart; i < str.size() && str.size() - i >= pattern.size() - pstart; ++i)
        {
            string temp = str.substr(sstart, i - sstart + 1);
            if (table.find(temp) != table.end())
            {
                if (table[temp] != pattern[pstart])
                {
                    continue;
                }
                else if (dfs(pattern, pstart + 1, str, i + 1))
                {
                    return true;
                }
            }
            else if (table.find(temp) == table.end() && chars.find(pattern[pstart]) != chars.end())
            {
                continue;
            }
            else
            {
                table[temp] = pattern[pstart];
                chars.insert(pattern[pstart]);
                if (dfs(pattern, pstart + 1, str, i + 1))
                {
                    return true;
                }
                else
                {
                    table.erase(temp);
                    chars.erase(pattern[pstart]);
                }
            }
        }
        
        return false;
    }
};

File: 0201-0300/0291-Word Pattern II/0291-Word Pattern II.py
class Solution:
    def wordPatternMatch(self, pattern: str, str: str) -> bool:
        table = {}
        mapped = set()
        def dfs(pstart, sstart):
            if pstart == len(pattern):
                return sstart == len(str)
            for i in range(sstart, len(str)):
                if len(str) - i < len(pattern) - pstart:
                    break
                temp = str[sstart : i + 1]
                if temp in table:
                    if table[temp] != pattern[pstart]:
                        continue
                    elif dfs(pstart + 1, i + 1):
                        return True
                elif pattern[pstart] not in mapped:
                    table[temp] = pattern[pstart]
                    mapped.add(pattern[pstart])
                    if dfs(pstart + 1, i + 1):
                        return True
                    del table[temp]
                    mapped.remove(pattern[pstart])
            
            return False
        return dfs(0, 0)
